### PR TITLE
Make copy_ex compatible with fluentd 0.14

### PR DESCRIPTION
**Pull request by sullerandras**
The current code does not seem to be compatible with 0.14, even though it is tagged with `v14`.

### CODE REVIEW

**Comment by sonots** [comment]
I thought it worked with v0.14 via its v0.12 compatible layer.
Did not it work?

**Comment by sullerandras** [comment]
@sonots It did not work for me unfortunately, the `emit` method in your plugin was never called by fluentd 0.14.23.

**Comment by sonots** [comment]
okay, I will check it. I want to make this plugin working with both v0.12 and v0.14.

**Comment by sonots** [comment]
Well, it worked with v0.14.

```
2017-11-24 17:17:01 +0900 [info]: parsing config file is succeeded path="example/fluent.conf"
2017-11-24 17:17:01 +0900 [warn]: 'type' is deprecated parameter name. use '@type' instead.
2017-11-24 17:17:01 +0900 [warn]: 'type' is deprecated parameter name. use '@type' instead.
2017-11-24 17:17:01 +0900 [warn]: 'type' is deprecated parameter name. use '@type' instead.
2017-11-24 17:17:01 +0900 [warn]: 'type' is deprecated parameter name. use '@type' instead.
2017-11-24 17:17:01 +0900 [warn]: both of Plugin @id and path for <storage> are not specified. Using on-memory store.
2017-11-24 17:17:01 +0900 [warn]: both of Plugin @id and path for <storage> are not specified. Using on-memory store.
2017-11-24 17:17:01 +0900 [info]: using configuration file: <ROOT>
  <source>
    type dummy
    tag "test"
  </source>
  <match **>
    type copy_ex
    <store ignore_error>
      type stdout
    </store>
    <store ignore_error>
      type stdout
    </store>
  </match>
</ROOT>
2017-11-24 17:17:01 +0900 [info]: starting fluentd-0.14.23 pid=21276 ruby="2.4.1"
2017-11-24 17:17:01 +0900 [info]: spawn command to main:  cmdline=["/Users/seo.naotoshi/.rbenv/versions/2.4.1/bin/ruby", "-Eascii-8bit:ascii-8bit", "-rbundler/setup", "/Users/seo.naotoshi/src/github.com/sonots/fluent-plugin-copy_ex/vendor/bundle/ruby/2.4.0/bin/fluentd", "-c", "example/fluent.conf", "--under-supervisor"]
2017-11-24 17:17:02 +0900 [info]: gem 'fluentd' version '0.14.23'
2017-11-24 17:17:02 +0900 [info]: gem 'fluent-plugin-copy_ex' version '0.0.3'
2017-11-24 17:17:02 +0900 [warn]: #0 'type' is deprecated parameter name. use '@type' instead.
2017-11-24 17:17:02 +0900 [info]: adding match pattern="**" type="copy_ex"
2017-11-24 17:17:02 +0900 [warn]: #0 'type' is deprecated parameter name. use '@type' instead.
2017-11-24 17:17:02 +0900 [warn]: #0 'type' is deprecated parameter name. use '@type' instead.
2017-11-24 17:17:02 +0900 [warn]: #0 'type' is deprecated parameter name. use '@type' instead.
2017-11-24 17:17:02 +0900 [info]: adding source type="dummy"
2017-11-24 17:17:02 +0900 [warn]: #0 both of Plugin @id and path for <storage> are not specified. Using on-memory store.
2017-11-24 17:17:02 +0900 [warn]: #0 both of Plugin @id and path for <storage> are not specified. Using on-memory store.
2017-11-24 17:17:02 +0900 [info]: #0 starting fluentd worker pid=21292 ppid=21276 worker=0
2017-11-24 17:17:02 +0900 [info]: #0 fluentd worker is now running worker=0
2017-11-24 17:17:02.288149000 +0900 fluent.info: {"worker":0,"message":"fluentd worker is now running worker=0"}
2017-11-24 17:17:02.288149000 +0900 fluent.info: {"worker":0,"message":"fluentd worker is now running worker=0"}
2017-11-24 17:17:03.019680000 +0900 test: {"message":"dummy"}
2017-11-24 17:17:03.019680000 +0900 test: {"message":"dummy"}
2017-11-24 17:17:04.065184000 +0900 test: {"message":"dummy"}
2017-11-24 17:17:04.065184000 +0900 test: {"message":"dummy"}
2017-11-24 17:17:05.092760000 +0900 test: {"message":"dummy"}
2017-11-24 17:17:05.092760000 +0900 test: {"message":"dummy"}
2017-11-24 17:17:06.028548000 +0900 test: {"message":"dummy"}
2017-11-24 17:17:06.028548000 +0900 test: {"message":"dummy"}
2017-11-24 17:17:07.060279000 +0900 test: {"message":"dummy"}
2017-11-24 17:17:07.060279000 +0900 test: {"message":"dummy"}
2017-11-24 17:17:08.100921000 +0900 test: {"message":"dummy"}
2017-11-24 17:17:08.100921000 +0900 test: {"message":"dummy"}
```

**Comment by sullerandras** [comment]
Here is an example fluent config file what reproduces the issue i was facing:
```
<source>
  @type "tail"
  format "json"
  path "input.log"
  pos_file "input.log.pos"
  tag "test.*"
  refresh_interval 1
</source>
<match **>
  @type copy_ex
  <store ignore_error>
    @type file

    append true
    path "output.log"
    time_slice_format test-%Y-%m-%d
    <format>
      @type json
      localtime false
    </format>
    <buffer>
      @type memory
      flush_mode immediate
      flush_at_shutdown true
    </buffer>
  </store>
</match>
```
Gemfile:
```ruby
source 'https://rubygems.org'

gem 'fluentd', '~> 0.14.23'
gem 'fluent-plugin-copy_ex', '~> 0.0.3'#, git: 'https://github.com/sullerandras/fluent-plugin-copy_ex.git', branch: 'fluentd-v0.14'
```
In the real config i use a different output plugin, not "file", but this demonstrates that the `output.log` file is not created with your version of `copy_ex` but it works with my fix.

**Comment by sonots** [comment]
Thanks, I could reproduce your problem.

It seemed ways to shutdown need to be fixed.

```
2017-11-24 17:30:24 +0900 [warn]: #0 thread doesn't exit correctly (killed or other reason) plugin=Fluent::Plugin::FileOutput title=:enqueue_thread thread=#<Thread:0x007ff11302ec28@/Users/seo.naotoshi/src/github.com/sonots/fluent-plugin-copy_ex/vendor/bundle/ruby/2.4.0/gems
/fluentd-0.14.23/lib/fluent/plugin_helper/thread.rb:70 aborting> error=nil
2017-11-24 17:30:24 +0900 [warn]: #0 thread doesn't exit correctly (killed or other reason) plugin=Fluent::Plugin::FileOutput title=:flush_thread_0 thread=#<Thread:0x007ff11302f100@/Users/seo.naotoshi/src/github.com/sonots/fluent-plugin-copy_ex/vendor/bundle/ruby/2.4.0/gems
/fluentd-0.14.23/lib/fluent/plugin_helper/thread.rb:70 aborting> error=nil
```

File: lib/fluent/plugin/out_copy_ex.rb [hunk 1]
```diff
@@ -1,71 +1,43 @@
...
         es = m
       end
 
-      # Here, we do not use OutputChain for custom
-      @outputs.each_index do |idx|
+      outputs.each_with_index do |output, idx|
         _es = @deep_copy ? es.dup : es
         begin
-          @emit_procs[idx].call(@outputs[idx], tag, _es, NullOutputChain.instance)
+          output.emit_events(tag, @deep_copy ? es.dup : es)
```
Comment (sonots) [inline lib/fluent/plugin/out_copy_ex.rb]: `_es` is available.

File: lib/fluent/plugin/out_copy_ex.rb [hunk 2]
```diff
@@ -1,71 +1,43 @@
-require 'fluent/output'
+require 'fluent/plugin/multi_output'
+require 'fluent/config/error'
+require 'fluent/event'
 
-module Fluent
+module Fluent::Plugin
```
Comment (sonots) [inline lib/fluent/plugin/out_copy_ex.rb:5]: Do we need to change the namespace here? If this is not necessary, could you keep as is?

**Comment by sonots** [comment]
This branch did not work with v0.12.41.

File: lib/fluent/plugin/out_copy_ex.rb (hunk 2 above)
Comment (sullerandras) [reply lib/fluent/plugin/out_copy_ex.rb:5]: I don't know. I copied the out_copy.rb file from fluentd: https://github.com/fluent/fluentd/blob/e2e0f64e006605cb0d47e25572913af857cda082/lib/fluent/plugin/out_copy.rb#L21

**Comment by sonots** [comment]
I think we can drop v0.10 support now.

**Comment by sullerandras** [comment]
I don't know why your plugin doesn't work with 0.14, [the plugin upgrade guide](https://docs.fluentd.org/v0.14/articles/plugin-update-from-v12) clearly states that 0.12 plugins should work with 0.14:
> Plugins which uses v0.12 API will be supported between Fluentd v0.14 and v1 (will be obsoleted at v2).

**Comment by sonots** [comment]
I guess we are hitting something outside of v0.12 compatibility layer.

**Comment by sullerandras** [comment]
I couldn't make it compatible with both 0.12 and 0.14. Maybe it's not possible for `MultiOutput` plugins?
So I guess support for fluentd < 0.14 needs to be dropped. Shall i update the gemspec as it is specified in the [upgrade guide](https://docs.fluentd.org/v0.14/articles/plugin-update-from-v12)?
```
gem.add_runtime_dependency "fluentd", ">= 0.14.0", "< 2"
```

**Comment by sonots** [comment]
I've fixed via https://github.com/sonots/fluent-plugin-copy_ex/pull/5. Thanks anyway!

**Pull request closed without merging**